lcls-twincat-motion/Library/POUs/Motion/FB_MotionPneumaticActuator.TcPOU

```diff
@@ -0,0 +1,209 @@
+(*PMPS PERMISSION*)
+// yet to be implemented
+
+(* Can't have bRetract_SW and  bInsert_SW both be true*)
+If (stPneumaticActuator.bRetract_SW) AND (stPneumaticActuator.bInsert_SW) THEN
+    stPneumaticActuator.bRetract_SW := FALSE;
+    stPneumaticActuator.bInsert_SW := FALSE;
+END_IF
+//Redundant??
+(*Check if both digital outputs active at the same time, and clear all*)
+IF stPneumaticActuator.q_bInsert THEN
+    stPneumaticActuator.q_bRetract := FALSE;
+    stPneumaticActuator.bRetract_SW:= FALSE;
+END_IF;
+IF stPneumaticActuator.q_bRetract THEN
+    stPneumaticActuator.q_bInsert := FALSE;
+    stPneumaticActuator.bInsert_SW:= FALSE;
+END_IF;
+
+(*Actuate the device*)
+stPneumaticActuator.q_bRetract := stPneumaticActuator.bRetractOK AND stPneumaticActuator.bRetract_SW AND NOT stPneumaticActuator.bInsert_SW ;
+stPneumaticActuator.q_bInsert := stPneumaticActuator.bInsertOK AND stPneumaticActuator.bInsert_SW AND NOT stPneumaticActuator.bRetract_SW ;
+
+(*Reset the Control signal when command has been executed and give time to ensure the actuator is fully seated in either direction*)
+IF (NOT ibSingleCntrl AND NOT ibCntrlHold) THEN
+   IF (stPneumaticActuator.bRetract_SW AND stPneumaticActuator.i_bOutLimitSwitch AND tRetractLimitSwitch.Q ) THEN stPneumaticActuator.q_bRetract := FALSE; END_IF
+   IF (stPneumaticActuator.bInsert_SW AND stPneumaticActuator.i_bInLimitSwitch AND tInsertLimitSwitch.Q) THEN stPneumaticActuator.q_bInsert := FALSE; END_IF
+END_IF
+
+(*Timers*)
+tInserttimeout(IN:= stPneumaticActuator.q_bInsert, PT := tTimeOutDuration );
+tRetracttimeout(IN:= stPneumaticActuator.q_bRetract, PT := tTimeOutDuration);
+tInsertLimitSwitch(IN:= stPneumaticActuator.i_bInLimitSwitch, PT := tLimitSwitchLatchDuration);
+tRetractLimitSwitch(IN:= stPneumaticActuator.i_bOutLimitSwitch, PT := tLimitSwitchLatchDuration);
+
+
+///Check moving postion timout
+IF NOT stPneumaticActuator.i_bInLimitSwitch AND tInserttimeout.Q THEN
+    stPneumaticActuator.bError := TRUE;
+    stPneumaticActuator.sErrorMessage:= 'Actuator insert timeout';
+ELSIF NOT stPneumaticActuator.i_bOutLimitSwitch AND tRetracttimeout.Q THEN
+    stPneumaticActuator.bError := TRUE;
+    stPneumaticActuator.sErrorMessage:= 'Actuator retract timeout';
+END_IF
+// Reset error
+stPneumaticActuator.bError R= stPneumaticActuator.bReset;
+
+(*FAST FAULT*)
+fbFF(i_xOK := xMPS_OK,
+    i_xReset := i_xReset,
+    io_fbFFHWO := io_fbFFHWO);
+
+(*Soft IO Mapping to Epics pvs*)
+ACT_IO();]]></ST>
+    </Implementation>
+    <Action Name="ACT_IO" Id="{090f58e5-b160-4478-a84f-328a90ba6204}">
+      <Implementation>
+        <ST><![CDATA[(*Inputs*)
+stPneumaticActuator.i_bInLimitSwitch :=  i_xInsertedLS;
+stPneumaticActuator.i_bOutLimitSwitch := i_xRetractedLS;
+
+(*outputs*)
+q_xInsert_DO:=stPneumaticActuator.q_bInsert;
+q_xRetract_DO:=stPneumaticActuator.q_bRetract;
```

Comment:
To clarify: for double-acting pneumatics, the engineer would link both of these and set the options appropriately? And for single-acting, the engineer would link one or the other based on the polarity of the pneumatic and set the hold option? e.g. a pneumatic that goes out when unpowered would need to link `q_xInsert_DO`?

Reply:
The set hold option, only applies to the double acting.
With single acting the hold option should always be true, otherwise every time you loose the signal the actuator would move to the other position. 
And yes the e.g is correct.